wkcuber/api/Layer.py

```diff
@@ -88,6 +96,131 @@ def set_bounding_box_size(self, size: Tuple[int, int, int]):
+                if buffer_width > 0:
+                    padding_shape = shape
+                    padding_shape[i+1] = buffer_width
+                    mag.write(data=np.zeros(padding_shape, dtype=mag.get_dtype()), offset=aligned_offset)
+                # pad right / bottom / back
+                buffer_width = ((aligned_offset + aligned_size) - (np.array(current_offset) + np.array(current_size)))[i]
+                if buffer_width > 0:
+                    padding_shape = shape
+                    padding_shape[i+1] = buffer_width
+                    right_offset = aligned_offset
+                    right_offset[i] = current_offset[i] + current_size[i]
+                    mag.write(data=np.zeros(padding_shape, dtype=mag.get_dtype()), offset=right_offset)
+
+        interpolation_mode = parse_interpolation_mode(interpolation_mode, self.name)
+        prev_mag = from_mag
+        target_mag = get_next_mag(prev_mag, scale)
+
+        while target_mag <= max_mag:
+            assert prev_mag < target_mag
+            assert target_mag.to_layer_name() not in self.mags
+
+            prev_mag_ds = self.mags[prev_mag.to_layer_name()]
+
+            mag_factors = [
+                t // s for (t, s) in zip(target_mag.to_array(), prev_mag.to_array())
+            ]
+
+            # initialize the new mag
+            target_mag_ds = self._initialize_mag_from_other_mag(target_mag, prev_mag_ds, compress)
+
+            # Get target view (aligned with mag_factors)
+            true_target_offset = target_mag_ds.get_view().global_offset
+            aligned_target_offset = mag_factors * (np.array(true_target_offset) // mag_factors)
+            target_mag_view = target_mag_ds.get_view(offset=aligned_target_offset, is_bounded=False)
+
+            # perform downsampling
+            with get_executor_for_args(args) as executor:
+                voxel_count_per_cube = np.prod(prev_mag_ds._get_file_dimensions())
+                job_count_per_log = math.ceil(1024 ** 3 / voxel_count_per_cube)  # log every gigavoxel of processed data
+
+                if buffer_edge_len is None:
+                    buffer_edge_len = determine_buffer_edge_len(prev_mag_ds.view) # DEFAULT_EDGE_LEN
+                job_args = (
+                    mag_factors,
+                    interpolation_mode,
+                    buffer_edge_len,
+                    compress,
+                    target_mag_ds._get_file_dimensions(),
+                    job_count_per_log,
+                )
+                prev_mag_ds.get_view().for_zipped_chunks(
+                    # this view is restricted to the bounding box specified in the properties
+                    downsample_cube_job,
+                    target_view=target_mag_view,
+                    job_args_per_chunk=job_args,
+                    source_chunk_size=np.array(target_mag_ds._get_file_dimensions()) * mag_factors,
+                    target_chunk_size=target_mag_ds._get_file_dimensions(),
```

Comment:
The idea is to use the file size of the target mag as a reference because concurrent writes need to be avoided. However, multiple of this would also be possible. For example the file dimension of a tiff is only (x, y, 1).